File: scripts/extract.py
#!/usr/bin/env python3
"""Extract clean content from Claude Code session transcripts.

Reads JSONL transcript files and outputs clean markdown with user messages
and short assistant responses. Supports finding unprocessed sessions.

Usage:
    python3 extract.py <transcript.jsonl>           # Extract one session
    python3 extract.py --unprocessed                 # List unprocessed sessions
    python3 extract.py --unprocessed --project <name> # Filter by project name
"""

import json
import os
import sys
from datetime import datetime
from pathlib import Path


def get_brain_root():
    """Resolve brain root from config. Search order:
    1. ~/.claude/plugin-config/remember/config.json  (user scope, persistent)
    2. .claude/plugin-config/remember/config.json    (project scope, persistent)
    3. ${PLUGIN_ROOT}/config.defaults.json           (shipped default)
    4. Hardcoded ~/remember
    """
    plugin_dir = Path(__file__).resolve().parent.parent
    config_candidates = [
        Path.home() / ".claude" / "plugin-config" / "remember" / "config.json",
        Path(".claude") / "plugin-config" / "remember" / "config.json",
        plugin_dir / "config.defaults.json",
    ]
    for config_path in config_candidates:
        if not config_path.exists():
            continue
        try:
            with open(config_path, encoding="utf-8") as f:
                config = json.load(f)
            data_root = config.get("paths", {}).get("data_root", "")
            if data_root:
                return Path(os.path.expanduser(data_root))
        except (json.JSONDecodeError, OSError):
            continue
    return Path.home() / "remember"


BRAIN_ROOT = get_brain_root()
PROCESSED_FILE = BRAIN_ROOT / ".processed_sessions"
CLAUDE_PROJECTS_DIR = Path.home() / ".claude" / "projects"
MAX_ASSISTANT_TEXT_LEN = 500


def get_processed_sessions():
    """Read set of already-processed session IDs."""
    if not PROCESSED_FILE.exists():
        return set()
    return set(PROCESSED_FILE.read_text(encoding="utf-8").strip().splitlines())


def mark_session_processed(session_id):
    """Append session ID to processed file."""
    PROCESSED_FILE.parent.mkdir(parents=True, exist_ok=True)
    with open(PROCESSED_FILE, "a", encoding="utf-8") as f:
        f.write(session_id + "\n")


def find_all_transcripts():
    """Find all JSONL transcript files across all projects."""
    transcripts = []
    if not CLAUDE_PROJECTS_DIR.exists():
        return transcripts
    for project_dir in CLAUDE_PROJECTS_DIR.iterdir():
        if not project_dir.is_dir():
            continue
        for jsonl in project_dir.glob("*.jsonl"):
            transcripts.append(jsonl)
    return transcripts


def get_session_id(path):
    """Extract session ID from transcript filename."""
    return Path(path).stem


def get_project_name(path):
    """Extract human-readable project name from the project directory."""
    parent = Path(path).parent.name
    parts = parent.split("-")
    # Find "projects" marker and take everything after
    try:
        idx = parts.index("projects")
        project_parts = [p for p in parts[idx + 1:] if p]
        if project_parts:
            return "/".join(project_parts)
    except ValueError:
        pass
    # No "projects" in path or no parts after it — use last 2 meaningful parts
    meaningful = [p for p in parts if p]
    return "/".join(meaningful[-2:]) if len(meaningful) >= 2 else parent


def is_noise_content(content):
    """Check if user message content is noise (commands, system tags, etc)."""
    if not isinstance(content, str):
        return False
    noise_prefixes = [
        "<local-command-",
        "<command-name>",
        "<system-",
        "<user-prompt-submit-hook>",
    ]
    return any(content.strip().startswith(p) for p in noise_prefixes)


def extract_user_text(message):
    """Extract clean text from a user message."""
    content = message.get("content", "")
    if isinstance(content, str):
        if is_noise_content(content):
            return None
        text = content.strip()
        return text if text else None
    elif isinstance(content, list):
        texts = []
        for block in content:
            if isinstance(block, dict) and block.get("type") == "text":
                t = block["text"].strip()
                if t and not is_noise_content(t):
                    texts.append(t)
            elif isinstance(block, str):
                if not is_noise_content(block):
                    texts.append(block.strip())
        return "\n".join(texts) if texts else None
    return None


def extract_assistant_text(message):
    """Extract short text responses from assistant messages (skip tool calls)."""
    content = message.get("content", [])
    if isinstance(content, str):
        return content.strip() if len(content) <= MAX_ASSISTANT_TEXT_LEN else None
    if not isinstance(content, list):
        return None
    texts = []
    has_tool_use = False
    for block in content:
        if not isinstance(block, dict):
            continue
        if block.get("type") == "tool_use":
            has_tool_use = True
        elif block.get("type") == "text":
            t = block.get("text", "").strip()
            if t:
                texts.append(t)

    if not texts:
        return None

    combined = "\n".join(texts)
    # Skip long assistant messages - they're usually code/tool output
    if len(combined) > MAX_ASSISTANT_TEXT_LEN:
        return None
    # If it was only tool calls with a tiny preamble, skip
    if has_tool_use and len(combined) < 20:
        return None
    return combined


def parse_timestamp(ts_str):
    """Parse ISO timestamp to datetime object."""
    try:
        return datetime.fromisoformat(ts_str.replace("Z", "+00:00"))
    except (ValueError, AttributeError):
        return None


def format_timestamp(ts_str, include_date=False):
    """Format ISO timestamp to readable format."""
    dt = parse_timestamp(ts_str)
    if not dt:
        return ""
    if include_date:
        return dt.strftime("%Y-%m-%d %H:%M")
    return dt.strftime("%H:%M")


def extract_session(path):
    """Extract clean content from a single transcript file.

    Returns dict with session metadata and extracted messages.
    """
    session_id = get_session_id(path)
    project = get_project_name(path)
    messages = []
    session_cwd = None
    first_ts = None
    last_ts = None

    with open(path, encoding="utf-8") as f:
        for line in f:
            try:
                entry = json.loads(line)
            except json.JSONDecodeError:
                continue

            entry_type = entry.get("type")
            ts = entry.get("timestamp", "")

            if first_ts is None and ts:
                first_ts = ts
            if ts:
                last_ts = ts

            if not session_cwd and entry.get("cwd"):
                session_cwd = entry["cwd"]

            if entry_type == "user":
                # Skip meta messages
                if entry.get("isMeta"):
                    continue
                msg = entry.get("message", {})
                text = extract_user_text(msg)
                if text:
                    messages.append({
                        "role": "user",
                        "text": text,
                        "time": format_timestamp(ts),
                        "timestamp": ts,
                    })

            elif entry_type == "assistant":
                msg = entry.get("message", {})
                text = extract_assistant_text(msg)
                if text:
                    messages.append({
                        "role": "assistant",
                        "text": text,
                        "time": format_timestamp(ts),
                        "timestamp": ts,
                    })

    # Detect if session spans multiple days
    spans_multiple_days = False
    if first_ts and last_ts:
        first_dt = parse_timestamp(first_ts)
        last_dt = parse_timestamp(last_ts)
        if first_dt and last_dt and first_dt.date() != last_dt.date():
            spans_multiple_days = True

    return {
        "session_id": session_id,
        "project": project,
        "cwd": session_cwd,
        "first_ts": first_ts,
        "last_ts": last_ts,
        "spans_multiple_days": spans_multiple_days,
        "session_date": first_ts,
        "messages": messages,
    }


def format_session_markdown(session):
    """Format extracted session data as clean markdown."""
    lines = []
    date_str = ""
    end_date_str = ""
    if session["first_ts"]:
        dt = parse_timestamp(session["first_ts"])
        if dt:
            date_str = dt.strftime("%Y-%m-%d")
    if session["last_ts"]:
        dt = parse_timestamp(session["last_ts"])
        if dt:
            end_date_str = dt.strftime("%Y-%m-%d")

    spans = session.get("spans_multiple_days", False)

    lines.append(f"# Session: {session['project']}")
    if date_str:
        if spans and end_date_str and date_str != end_date_str:
            lines.append(f"**Date:** {date_str} to {end_date_str}")
        else:
            lines.append(f"**Date:** {date_str}")
    lines.append(f"**Session date (use for journal/tasks):** {date_str}")
    if session["cwd"]:
        lines.append(f"**Working dir:** `{session['cwd']}`")
    lines.append(f"**Session ID:** `{session['session_id']}`")
    lines.append("")

    if not session["messages"]:
        lines.append("*(empty session)*")
        return "\n".join(lines)

    current_day = None
    for msg in session["messages"]:
        # Show date headers when session spans multiple days
        if spans and msg.get("timestamp"):
            msg_dt = parse_timestamp(msg["timestamp"])
            if msg_dt:
                msg_day = msg_dt.strftime("%Y-%m-%d")
                if msg_day != current_day:
                    current_day = msg_day
                    lines.append(f"### {msg_day}")
                    lines.append("")

        time_prefix = f"[{msg['time']}] " if msg["time"] else ""
        if msg["role"] == "user":
            lines.append(f"**{time_prefix}User:** {msg['text']}")
        else:
            lines.append(f"*{time_prefix}Claude: {msg['text']}*")
        lines.append("")

    return "\n".join(lines)


def cmd_extract(path):
    """Extract and print a single session."""
    if not os.path.exists(path):
        print(f"Error: file not found: {path}", file=sys.stderr)
        sys.exit(1)
    session = extract_session(path)
    print(format_session_markdown(session))


def cmd_list_unprocessed(project_filter=None):
    """List unprocessed transcript files."""
    processed = get_processed_sessions()
    transcripts = find_all_transcripts()

    unprocessed = []
    for t in transcripts:
        sid = get_session_id(t)
        if sid in processed:
            continue
        project = get_project_name(t)
        if project_filter and project_filter.lower() not in project.lower():
            continue
        # Get file size to skip tiny sessions
        size = t.stat().st_size
        if size < 500:  # Skip very small files (likely empty/init sessions)
            continue
        unprocessed.append((t, project, size))

    # Sort by modification time (oldest first)
    unprocessed.sort(key=lambda x: x[0].stat().st_mtime)

    if not unprocessed:
        print("No unprocessed sessions found.")
        return

    print(f"Found {len(unprocessed)} unprocessed session(s):\n")
    for path, project, size in unprocessed:
        sid = get_session_id(path)
        mtime = datetime.fromtimestamp(path.stat().st_mtime).strftime("%Y-%m-%d %H:%M")
        size_kb = size / 1024
        print(f"  {sid}")
        print(f"    Project: {project}")
        print(f"    Modified: {mtime}")
        print(f"    Size: {size_kb:.0f}KB")
        print(f"    Path: {path}")
        print()


def cmd_mark_processed(session_id):
    """Mark a session as processed."""
    mark_session_processed(session_id)
    print(f"Marked as processed: {session_id}")


def main():
    if len(sys.argv) < 2:
        print(__doc__)
        sys.exit(1)

    arg = sys.argv[1]

    if arg == "--unprocessed":
        project_filter = None
        if "--project" in sys.argv:
            idx = sys.argv.index("--project")
            if idx + 1 < len(sys.argv):
                project_filter = sys.argv[idx + 1]
        cmd_list_unprocessed(project_filter)
    elif arg == "--mark-processed":
        if len(sys.argv) < 3:
            print("Usage: extract.py --mark-processed <session-id>", file=sys.stderr)
            sys.exit(1)
        cmd_mark_processed(sys.argv[2])
    else:
        cmd_extract(arg)


if __name__ == "__main__":
    main()

File: scripts/extract.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');
const os = require('os');
const { getBrainRoot, loadConfig } = require('./config');

const BRAIN_ROOT = getBrainRoot();
const PROCESSED_FILE = path.join(BRAIN_ROOT, '.processed_sessions');
const CLAUDE_PROJECTS_DIR = path.join(os.homedir(), '.claude', 'projects');
const MAX_ASSISTANT_TEXT_LEN = loadConfig().extract?.max_assistant_text_len || 500;
const MIN_SESSION_SIZE = loadConfig().extract?.min_session_size || 500;

// --- Session tracking ---

function getProcessedSessions() {
  try {
    return new Set(fs.readFileSync(PROCESSED_FILE, 'utf-8').trim().split('\n').filter(Boolean));
  } catch {
    return new Set();
  }
}

function markSessionProcessed(sessionId) {
  fs.mkdirSync(path.dirname(PROCESSED_FILE), { recursive: true });
  fs.appendFileSync(PROCESSED_FILE, sessionId + '\n', 'utf-8');
}

// --- Transcript discovery ---

function findAllTranscripts() {
  const transcripts = [];
  if (!fs.existsSync(CLAUDE_PROJECTS_DIR)) return transcripts;
  for (const projDir of fs.readdirSync(CLAUDE_PROJECTS_DIR)) {
    const projPath = path.join(CLAUDE_PROJECTS_DIR, projDir);
    let stat;
    try { stat = fs.statSync(projPath); } catch { continue; }
    if (!stat.isDirectory()) continue;
    for (const file of fs.readdirSync(projPath)) {
      if (file.endsWith('.jsonl')) {
        transcripts.push(path.join(projPath, file));
      }
    }
  }
  return transcripts;
}

function getSessionId(filePath) {
  return path.basename(filePath, '.jsonl');
}

function getProjectName(filePath) {
  const parent = path.basename(path.dirname(filePath));
  const parts = parent.split('-');
  const idx = parts.indexOf('projects');
  if (idx !== -1) {
    const projectParts = parts.slice(idx + 1).filter(Boolean);
    if (projectParts.length) return projectParts.join('/');
  }
  const meaningful = parts.filter(Boolean);
  return meaningful.length >= 2 ? meaningful.slice(-2).join('/') : parent;
}

// --- Content extraction ---

const NOISE_PREFIXES = [
  '<local-command-',
  '<command-name>',
  '<system-',
  '<user-prompt-submit-hook>',
];

function isNoiseContent(content) {
  if (typeof content !== 'string') return false;
  const trimmed = content.trim();
  return NOISE_PREFIXES.some(p => trimmed.startsWith(p));
}

function extractUserText(message) {
  const content = message.content;
  if (typeof content === 'string') {
    if (isNoiseContent(content)) return null;
    const text = content.trim();
    return text || null;
  }
  if (Array.isArray(content)) {
    const texts = [];
    for (const block of content) {
      if (typeof block === 'object' && block?.type === 'text') {
        const t = block.text?.trim();
        if (t && !isNoiseContent(t)) texts.push(t);
      } else if (typeof block === 'string') {
        if (!isNoiseContent(block)) texts.push(block.trim());
      }
    }
    return texts.length ? texts.join('\n') : null;
  }
  return null;
}

function extractAssistantText(message) {
  const content = message.content;
  if (typeof content === 'string') {
    return content.length <= MAX_ASSISTANT_TEXT_LEN ? content.trim() || null : null;
  }
  if (!Array.isArray(content)) return null;

  const texts = [];
  let hasToolUse = false;
  for (const block of content) {
    if (typeof block !== 'object' || !block) continue;
    if (block.type === 'tool_use') hasToolUse = true;
    else if (block.type === 'text') {
      const t = block.text?.trim();
      if (t) texts.push(t);
    }
  }
  if (!texts.length) return null;
  const combined = texts.join('\n');
  if (combined.length > MAX_ASSISTANT_TEXT_LEN) return null;
  if (hasToolUse && combined.length < 20) return null;
  return combined;
}

// --- Timestamp helpers ---

function parseTimestamp(tsStr) {
  if (!tsStr) return null;
  try {
    const d = new Date(tsStr);
    return isNaN(d.getTime()) ? null : d;
  } catch {
    return null;
  }
}

function formatTimestamp(tsStr, includeDate = false) {
  const dt = parseTimestamp(tsStr);
  if (!dt) return '';
  if (includeDate) {
    return dt.toISOString().slice(0, 16).replace('T', ' ');
  }
  return dt.toISOString().slice(11, 16);
}

function formatDateOnly(tsStr) {
  const dt = parseTimestamp(tsStr);
  return dt ? dt.toISOString().slice(0, 10) : '';
}

// --- Session extraction ---

function extractSession(filePath) {
  const sessionId = getSessionId(filePath);
  const project = getProjectName(filePath);
  const messages = [];
  let sessionCwd = null;
  let firstTs = null;
  let lastTs = null;

  const text = fs.readFileSync(filePath, 'utf-8');
  for (const line of text.split('\n')) {
    if (!line.trim()) continue;
    let entry;
    try { entry = JSON.parse(line); } catch { continue; }

    const ts = entry.timestamp || '';
    if (!firstTs && ts) firstTs = ts;
    if (ts) lastTs = ts;
    if (!sessionCwd && entry.cwd) sessionCwd = entry.cwd;

    if (entry.type === 'user') {
      if (entry.isMeta) continue;
      const msg = entry.message || {};
      const text = extractUserText(msg);
      if (text) {
        messages.push({ role: 'user', text, time: formatTimestamp(ts), timestamp: ts });
      }
    } else if (entry.type === 'assistant') {
      const msg = entry.message || {};
      const text = extractAssistantText(msg);
      if (text) {
        messages.push({ role: 'assistant', text, time: formatTimestamp(ts), timestamp: ts });
      }
    }
  }

  let spansMultipleDays = false;
  if (firstTs && lastTs) {
    const firstDate = formatDateOnly(firstTs);
    const lastDate = formatDateOnly(lastTs);
    if (firstDate && lastDate && firstDate !== lastDate) spansMultipleDays = true;
  }

  return {
    session_id: sessionId,
    project,
    cwd: sessionCwd,
    first_ts: firstTs,
    last_ts: lastTs,
    spans_multiple_days: spansMultipleDays,
    session_date: firstTs,
    messages,
  };
}

// --- Formatting ---

function formatSessionMarkdown(session) {
  const lines = [];
  const dateStr = formatDateOnly(session.first_ts);
  const endDateStr = formatDateOnly(session.last_ts);
  const spans = session.spans_multiple_days;

  lines.push(`# Session: ${session.project}`);
  if (dateStr) {
    lines.push(spans && endDateStr && dateStr !== endDateStr
      ? `**Date:** ${dateStr} to ${endDateStr}`
      : `**Date:** ${dateStr}`);
  }
  lines.push(`**Session date (use for journal/tasks):** ${dateStr}`);
  if (session.cwd) lines.push(`**Working dir:** \`${session.cwd}\``);
  lines.push(`**Session ID:** \`${session.session_id}\``);
  lines.push('');

  if (!session.messages.length) {
    lines.push('*(empty session)*');
    return lines.join('\n');
  }

  let currentDay = null;
  for (const msg of session.messages) {
    if (spans && msg.timestamp) {
      const msgDay = formatDateOnly(msg.timestamp);
      if (msgDay && msgDay !== currentDay) {
        currentDay = msgDay;
        lines.push(`### ${msgDay}`);
        lines.push('');
      }
    }
    const timePrefix = msg.time ? `[${msg.time}] ` : '';
    if (msg.role === 'user') {
      lines.push(`**${timePrefix}User:** ${msg.text}`);
    } else {
      lines.push(`*${timePrefix}Claude: ${msg.text}*`);
    }
    lines.push('');
  }

  return lines.join('\n');
}

// --- CLI commands ---

function cmdExtract(filePath) {
  if (!fs.existsSync(filePath)) {
    process.stderr.write(`Error: file not found: ${filePath}\n`);
    process.exit(1);
  }
  const session = extractSession(filePath);
  console.log(formatSessionMarkdown(session));
}

function cmdListUnprocessed(projectFilter) {
  const processed = getProcessedSessions();
  const transcripts = findAllTranscripts();

  const unprocessed = [];
  for (const t of transcripts) {
    const sid = getSessionId(t);
    if (processed.has(sid)) continue;
    const project = getProjectName(t);
    if (projectFilter && !project.toLowerCase().includes(projectFilter.toLowerCase())) continue;
    let size;
    try { size = fs.statSync(t).size; } catch { continue; }
    if (size < MIN_SESSION_SIZE) continue;
    unprocessed.push({ path: t, project, size });
  }

  unprocessed.sort((a, b) => {
    try {
      return fs.statSync(a.path).mtimeMs - fs.statSync(b.path).mtimeMs;
    } catch { return 0; }
  });

  if (!unprocessed.length) {
    console.log('No unprocessed sessions found.');
    return;
  }

  console.log(`Found ${unprocessed.length} unprocessed session(s):\n`);
  for (const u of unprocessed) {
    const sid = getSessionId(u.path);
    let mtime;
    try {
      const mt = fs.statSync(u.path).mtime;
      mtime = mt.toISOString().slice(0, 16).replace('T', ' ');
    } catch {
      mtime = 'unknown';
    }
    const sizeKb = (u.size / 1024).toFixed(0);
    console.log(`  ${sid}`);
    console.log(`    Project: ${u.project}`);
    console.log(`    Modified: ${mtime}`);
    console.log(`    Size: ${sizeKb}KB`);
    console.log(`    Path: ${u.path}`);
    console.log();
  }
}

function cmdMarkProcessed(sessionId) {
  markSessionProcessed(sessionId);
  console.log(`Marked as processed: ${sessionId}`);
}

// --- Main ---

function main() {
  const args = process.argv.slice(2);
  if (!args.length) {
    console.log('Usage:');
    console.log('  node extract.js <transcript.jsonl>');
    console.log('  node extract.js --unprocessed [--project <name>]');
    console.log('  node extract.js --mark-processed <session_id>');
    process.exit(1);
  }

  if (args[0] === '--unprocessed') {
    const projIdx = args.indexOf('--project');
    const projectFilter = projIdx !== -1 && args[projIdx + 1] ? args[projIdx + 1] : null;
    cmdListUnprocessed(projectFilter);
  } else if (args[0] === '--mark-processed') {
    if (!args[1]) {
      process.stderr.write('Usage: extract.js --mark-processed <session-id>\n');
      process.exit(1);
    }
    cmdMarkProcessed(args[1]);
  } else {
    cmdExtract(args[0]);
  }
}

main();
